src/sinks/file.rs

```diff
@@ -0,0 +1,316 @@
+            path_template,
+            encoding,
+            close_timeout_secs,
+            partitions: HashMap::new(),
+            last_accessed: HashMap::new(),
+            closing: Vec::new(),
+        }
+    }
+
+    fn collect_old_files(&mut self) {
+        let mut recently_outdated = vec![];
+        for timeout in self.close_timeout_secs {
+            self.last_accessed.retain(|path, time| {
+                if time.elapsed().as_secs() > timeout {
+                    debug!(message = "removing sink", file = ?path);
+                    recently_outdated.push(path.clone());
+                    false
+                } else {
+                    true
+                }
+            });
+        }
+
+        let mut recently_outdated = recently_outdated
+            .into_iter()
+            .map(|ref path| self.partitions.remove(path).unwrap())
+            .collect();
+
+        //it is easier to empty `closing` buffer and put back not closed sinks, see `poll_close`
+        let mut closing: Vec<EmbeddedFileSink> = self.closing.drain(..).collect();
+        closing.append(&mut recently_outdated);
+
+        for file in closing.into_iter() {
+            self.poll_close(file);
+        }
+    }
+
+    fn poll_close(&mut self, mut file: EmbeddedFileSink) {
+        match file.close() {
+            Err(err) => error!("Error while closing FileSink: {:?}", err),
+            Ok(Async::Ready(())) => debug!("a FileSink closed"),
+            Ok(Async::NotReady) => self.closing.push(file),
+        }
+    }
+}
+
+impl Sink for PartitionedFileSink {
+    type SinkItem = Event;
+    type SinkError = ();
+
+    fn start_send(&mut self, event: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+        match self.path_template.render(&event) {
+            Ok(bytes) => {
+                let path = PathBuf::from(String::from_utf8_lossy(&bytes).as_ref());
+
+                let partition = self
+                    .partitions
+                    .entry(path.clone())
+                    .or_insert(FileSink::new_with_encoding(&path, self.encoding.clone()));
+
+                self.last_accessed.insert(path.clone(), Instant::now());
+                partition.start_send(event)
+            }
+
+            Err(missing_keys) => {
+                warn!(
+                    message = "Keys do not exist on the event. Dropping event.",
+                    keys = ?missing_keys
+                );
+                Ok(AsyncSink::Ready)
+            }
+        }
+    }
+
+    fn poll_complete(&mut self) -> Result<Async<()>, Self::SinkError> {
+        self.partitions
+            .iter_mut()
+            .for_each(|(path, partition)| match partition.poll_complete() {
+                Ok(_) => {}
+                Err(()) => error!("Error in downstream FileSink with path {:?}", path),
+            });
+
+        self.collect_old_files();
+
+        let files: Vec<PathBuf> = self.partitions.keys().map(|path| path.clone()).collect();
```

Comment:
this clone seems really expensive for each call to poll_complete which could happen many times.  We should avoid this.